cantools/database/can/cpp_source.py

```diff
@@ -0,0 +1,1037 @@
+ *
+{comment}\
+ */
+class {database_message_name} : public Frame {{
+public:
+    /** Constructor */
+    {database_message_name}(uint8_t* buffer, size_t size);
+
+    {database_message_name}& GetMessage() {{ 
+        return *this; 
+    }}
+ 
+    uint8_t* buffer() {{
+        return _buffer;
+    }}
+ 
+    size_t buffer_size() {{
+        return _buffer_size;
+    }}
+  
+    // Clear buffer
+    void clear() {{
+        memset(_buffer, 0, _buffer_size);
+    }}
+
+{members}
+
+private:
+    uint8_t* _buffer;
+    size_t _buffer_size;
+
+{encode_decode_members}
+}};
+'''
+
+SIGNAL_DECLARATION_FMT = '''\
+    /**
+     * Signal {name}.
+{comment}\
+     * Range: {range}
+     * Scale: {scale}
+     * Offset: {offset}
+{additional_comments}\
+     */
+    double {name}();
+    {type_name} {name}_raw();
+    bool set_{name}(const double& value);
+    bool {name}_in_range(const double& value);
+    bool {name}_raw_in_range(const {type_name}& value) const;
+    std::string {name}_data_format() const;
+    {additional}\
+'''
+
+SIGNAL_ENCODE_DECODE_DECLARATION_FMT = '''\
+    {type_name} {name}_encode(double value);
+    double {name}_decode({type_name} value);
+'''
+
+MESSAGE_DEFINITION_FMT = '''\
+{constructor}
+{signal_definitions}
+'''
+
+MESSAGE_CONSTRUCTOR_DEFINITION_FMT = '''\
+{database_message_name}::{database_message_name}(uint8_t* buffer, size_t size)
+    : Frame({id}u, "{database_message_name}", {length}u, {extended}, {cycle_time}u)
+    , _buffer(buffer)
+    , _buffer_size(size)
+{{}}
+'''
+
+SIGNAL_DEFINITION_FMT = '''\
+double {database_message_name}::{name}() {{
+{contents}
+}}
+
+'''
+
+SIGNAL_DEFINITION_RAW_FMT = '''\
+{type_name} {database_message_name}::{name}_raw() {{
+{contents}
+}}
+
+'''
+
+SIGNAL_DEFINITION_SET_FMT = '''\
+bool {database_message_name}::set_{name}(const double& value) {{
+
+{contents}
+}}
+
+'''
+
+SIGNAL_DEFINITION_IN_RANGE_FMT = '''\
+bool {database_message_name}::{name}_in_range(const double& value) {{
+    {type_name} {snake_name} = {name}_encode(value);
+    return {name}_raw_in_range({snake_name});
+}}
+
+'''
+
+SIGNAL_DEFINITION_RAW_IN_RANGE_FMT = '''\
+bool {database_message_name}::{name}_raw_in_range(const {type_name}& value) const {{
+    return ({check});
+}}
+
+'''
+
+SIGNAL_DEFINITION_DATA_FORMAT_FMT = '''\
+std::string {database_message_name}::{name}_data_format() const {{
+    return "{data_format}";
+}}
+
+'''
+
+SIGNAL_DEFINITION_SPN_FMT = '''\
+uint32_t {database_message_name}::{name}_SPN() const {{
+    return {spn};
+}}
+
+'''
+
+SIGNAL_DEFINITION_ENCODE_DECODE_FMT = '''\
+{type_name} {database_message_name}::{name}_encode(double value) {{
+    return static_cast<{type_name}>({encode});
+}}
+
+double {database_message_name}::{name}_decode({type_name} value) {{
+    return ({decode});
+}}
+'''
+
+SIGN_EXTENSION_FMT = '''
+    if (({name} & (1{suffix} << {shift})) != 0{suffix}) {{
+        {name} |= 0x{mask:x}{suffix};
+    }}
+
+'''
+
+
+class Signal(object):
```

Comment:
Is this class substantially similar to the one in the c_source.py?  If so, maybe inheriting from it and modifying what you need would be cleaner than copying.  If it is very different, then disregard.

Reply:
Will remove duplication